docs/setup.md

```diff
@@ -2,7 +2,146 @@
+```
+
+If you have other dependencies, you should make a note of these, you'll be
+able to install them using the cibuildwheel options [`CIBW_BEFORE_BUILD`](options.md#before-build) or
+[`CIBW_BEFORE_ALL`](options.md#before-all).
+
+### Actions you need to perform before building
+
+You might need to run some other commands before building, like running a
+script that performs codegen or downloading some data that's not stored in
+your source tree. There are a couple ways to deal with this:
+
+-   Incorporate this into your build process, by adding it to your package's
+    `setup.py`. You can add extra build steps using a structure like this:
+
+    ```python
+    import subprocess
+    import setuptools
+    import setuptools.command.build_py
+
+
+    class BuildPyCommand(setuptools.command.build_py.build_py):
+      """Custom build command."""
+
+      def run(self):
+        # your custom build steps here
+        # e.g.
+        #   subprocess.run(['python', 'scripts/my_custom_script.py'], check=True)
+        setuptools.command.build_py.build_py.run(self)
+
+
+    setuptools.setup(
+        cmdclass={
+            'build_py': BuildPyCommand,
+        },
+        # Usual setup() args.
+        # ...
+    )
+    ```
+
+    This method is usually preferred because in addition to adding being
+    included in the wheel build process, it will help users building from
+    source tarballs as well.
+
+-   Alternatively, you can instruct cibuildwheel to run commands before
+    building your wheel. Take a look at the cibuildwheel options
+    [`CIBW_BEFORE_BUILD`](options.md#before-build) or
+    [`CIBW_BEFORE_ALL`](options.md#before-all).
+
+### Environment variables
+
+Your wheel build might need some environment variables to be set. Consider
+incorporating these into setup.py to allow source tarballs to build (for
+example, using [`extra_compile_args` or
+`extra_link_args`](https://docs.python.org/3/distutils/setupscript.html#other-options)),
+but otherwise, make a note of these for inclusion in the cibuildwheel option
+[`CIBW_ENVIRONMENT`](options.md#environment).
+
+# Run cibuildwheel locally
+
+If you've got [Docker](https://www.docker.com/products/docker-desktop)
+installed on your development machine, you can run a Linux build without even
+touching CI.
+
+!!! tip
+    Even Windows can run Linux containers these days, but there are a few
+    hoops to jump through. Check [this
+    document](https://docs.microsoft.com/en-us/virtualization/windowscontainers/quick-start/quick-start-windows-10-linux)
+    for more info.
+
+This is convenient as it's quicker to iterate, and because the builds are
+happening in manylinux Docker containers, they're perfectly reproducable.
+
+Install cibuildwheel and run a build like this:
+
+```sh
+pip install cibuildwheel
+cibuildwheel --platform linux
+```
+
+You should see the builds taking place. You can experiment with options by exporting environment variables, for example:
```

Comment:
I would mention the config mode here. It is _fantastic_ for local builds - in fact, I'd recommend (in general, maybe not right here) that the config should always allow a default build to work on any machine, with environment variables used to configure for a specific item in a CI matrix. That is, if you don't support pypy, I'd always add a `skip` item for it, even if in CI you never run a build where that matters by always setting `CIBW_BUILD`.

FYI, I would personally always run it this way:

```console
# CIBW_BUILD='cp39-*' pipx run cibuildwheel --platform linux
```

(the `tool.cibuildwheel.linux.before-all` setting would be in the `pyproject.toml` location)

Reply:
good point, i've added this